### PR TITLE
Add codedeploy 

### LINKED ISSUE #454: Support CodeDeploy Applications
It would be great if this tool supported nuking [CodeDeploy Applications](https://docs.aws.amazon.com/codedeploy/latest/userguide/applications.html).

### PR DESCRIPTION
**Pull request by robpickerill**
## Description

Fixes #454 

This PR introduces the ability to nuke CodeDeploy applications.

## TODOs

- [x] Add tests, I found myself writing a mock for AWS APIs, and had to step back as it was getting complex. I know you like to provision resources and delete them as integration testing, so let me know how you would like me to proceed with testing here.

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.
- [ ] Attention Grunts - if this PR adds support for a new resource, ensure the `nuke_sandbox` and `nuke_phxdevops` jobs in `.circleci/config.yml` have been updated with appropriate exclusions (either directly in the job or via the `.circleci/nuke_config.yml` file) to prevent nuking IAM roles, groups, resources, etc that are important for the test accounts.

## Release Notes (draft)

- Adds CodeDeploy support

### CODE REVIEW

**Comment by hongil0316** [comment]
Hey @robpickerill, thanks for raising the PR. Would you be able to write unit test for the `aws/codedeploy_application.go` file you created?

**Comment by hongil0316** [comment]
The overall code LGTM. Once you have the unit test, I can trigger the CI test pipeline. Thanks! :)

**Comment by robpickerill** [comment]
> The overall code LGTM. Once you have the unit test, I can trigger the CI test pipeline. Thanks! :)

Thanks, tests added, running locally all was working well. Your welcome to kick CI off now and I'll check for any problems.

File: aws/codedeploy_application_types.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+type CodeDeployApplications struct {
+	AppNames []*string
+}
+
+// ResourceName - the simple name of the aws resource
+func (c CodeDeployApplications) ResourceName() string {
+	return "codedeploy-application"
+}
+
+// ResourceIdentifiers - The instance ids of the ec2 instances
```
Comment (hongil0316) [inline aws/codedeploy_application_types.go:18]: nit: is this comment relevant here? Let's update

File: aws/codedeploy_application.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+// batchDescribeAndFilterCodeDeployApplications - Describe the CodeDeploy Applications and filter out the ones that should be excluded by CreationDate.
+func batchDescribeAndFilterCodeDeployApplications(session *session.Session, identifiers []string, excludeAfter time.Time) ([]string, error) {
+	svc := codedeploy.New(session)
+
+	applicationNames := []string{}
+
+	// BatchGetApplications can only take 100 identifiers at a time, so we have to break up the identifiers into chunks of 100.
+	for i := 0; i < len(identifiers); i += 100 {
+
+		// Determine the end index of the identifiers slice.
```
Comment (hongil0316) [inline aws/codedeploy_application.go:50]: Would it make sense to split the identifiers into chunks (w/ max size 100) and iterate through the chunks? I think that would have better readability. Reference: https://freshman.tech/snippets/go/split-slice-into-chunks/

Currently, it's a bit hard to read through the logic. WDYT?

**Review by hongil0316 — COMMENTED** [review]
Overall LGTM, left a few comments. Will trigger CI pipeline in the meaintime.

File: aws/codedeploy_application.go (hunk 2 above)
Comment (robpickerill) [reply aws/codedeploy_application.go]: I refactored this, would you take another look?

File: aws/codedeploy_application_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/codedeploy"
+	"github.com/gruntwork-io/cloud-nuke/config"
+)
+
+func randomString() string {
```
[the remark this answers is not in the document]
Comment (robpickerill) [reply aws/codedeploy_application_test.go:16]: Let's use random.UniqueId, I can drop randomString, leave with me briefly

File: aws/codedeploy_application.go (hunk 2 above)
Comment (hongil0316) [reply aws/codedeploy_application.go]: Sure. Let me know when it's ready for review again :)

**Comment by hongil0316** [comment]
It seems like there's some build failure, not sure if you were able to fix it with new commits

```
aws/aws.go:1698:39: cannot use applications (variable of type []string) as type []*string in assignment
aws/codedeploy_application_types.go:20:9: undefined: awsgo

```

**Comment by robpickerill** [comment]
>It seems like there's some build failure, not sure if you were able to fix it with new commits

Can you retry for me?

**Comment by hongil0316** [comment]
Triggered the test pipeline again

**Comment by hongil0316** [comment]
Test pipeline seems to be good. LGTM

**Comment by hongil0316** [comment]
Hey @robpickerill, are you making further changes or this is ready for final review?

**Comment by robpickerill** [comment]
Good to go, thanks!

**Review by ellisonc — APPROVED** [review]

**Pull request merged**